### ISSUE TITLE
The Rate Limiting feature doesn't work when use DistributedCache

### ISSUE THREAD

**Issue by rocgao** [ThreeMammals/Ocelot#609]
## Expected Behavior  
External http request should be blocked when quota exceeded.

## Actual Behavior
External http request was forwarded to downstream servers always.

## Steps to Reproduce the Problem

  1. Adding DistributedCache. Such as RedisCache, which is supplied by Microsoft.Extensions.Caching.Redis nuget package.
  1.  Enabling DistributedCache. Refer to following code:
``` c#
       // Startup.cs
       public void ConfigureServices(IServiceCollection services)
        {
            //使用基于Redis缓存的RateLimiting
            services.Configure<RedisCacheOptions>(Configuration.GetSection("RedisCacheOptions"))
                .AddSingleton<IDistributedCache,RedisCache>()
                .AddSingleton<IRateLimitCounterHandler,DistributedCacheRateLimitCounterHanlder>();

            services.AddOcelot();
        }
```   
  3.  Configuring RateLimiting options in Ocelot.json .
```json
      "ReRoutes": [
        {
            "DownstreamPathTemplate": "/health",
            "DownstreamScheme": "http",
            "DownstreamHostAndPorts": [
                {
                    "Host": "localhost",
                    "Port": 6001
                }
            ],
            "UpstreamPathTemplate": "/ocelot/health",
            "RateLimitOptions": {
                "ClientWhitelist": [],
                "EnableRateLimiting": true,
                "Period": "20s",
                "PeriodTimespan": 5,
                "Limit": 1
            }
        }
```
  4. Sent http request to Ocelot server, it will always response following content. Attention to the **X-Rate-Limit-Reset** header. 
```
HTTP/1.1 200 OK
Date: Tue, 11 Sep 2018 02:53:36 GMT
Content-Type: application/json
Server: Kestrel
Content-Length: 18
X-Rate-Limit-Limit: 20s
X-Rate-Limit-Remaining: 1
X-Rate-Limit-Reset: 0001-01-01T00:00:00.0000000Z
```

## Specifications

  - Version: 10.*
  - Platform: MacOS
  - Subsystem:

**Comment by rocgao** [comment]
I have found out the bug. The DistributedCacheRateLimitCounterHanlder.Get will be called, when the RedisCache is enabled. Within the method, call JsonConvert.DeserializeObject funtion to load RateLimitCounter instance from json string. But both Timestamp  and TotalRequests are readonly of the RateLimitCounter structor.  So to fix this bug, should apply the JsonConstructorAttribute to RateLimitCounter's constructor and rename totalRequest parameter to totalRequests.
```c#
       [JsonConstructor]
        public RateLimitCounter(DateTime timestamp, long totalRequests)
        {
            Timestamp = timestamp;
            TotalRequests = totalRequests;
        }
```

**Comment by TomPallister** [comment]
@rocgao thanks for this issue and the fix. I really appreciate it. I will merge this into NuGet package ASAP.

**Comment by TomPallister** [comment]
fixed in 11.0.3, any problems let me know

**Comment by raman-m** [comment]
@rocgao @geffzhang 
Are you still with Ocelot? 

Regarding the [user scenario](https://github.com/ThreeMammals/Ocelot/issues/609#issue-358872685) with RedisCache...
How did/does this setup work in Production env?
Are you satisfied? 
Can we turn this setup of services to real Ocelot feature?
We could develop at least new Ocelot DI-extension method for the first time.
